pkg/inventory/inventory-client.go

```diff
@@ -6,473 +6,402 @@ package inventory
 	}
-	return clusterInvObjects, err
-}
-
-func (cic *ClusterClient) ListClusterInventoryObjs(ctx context.Context) (map[string]object.ObjMetadataSet, error) {
-	// Define the mapping
-	mapping, err := cic.mapper.RESTMapping(cic.gvk.GroupKind(), cic.gvk.Version)
+	obj, err := cic.client.Namespace(soi.Namespace()).Get(ctx, soi.Name(), metav1.GetOptions{})
 	if err != nil {
 		return nil, err
 	}
+	return cic.fromUnstructured(obj)
+}
 
-	// retrieve the list from the cluster
-	clusterInvs, err := cic.dc.Resource(mapping.Resource).List(ctx, metav1.ListOptions{})
-	if err != nil && !apierrors.IsNotFound(err) {
+// List the in-cluster inventory
+// Used by the CLI commands
+func (cic *UnstructuredClient) List(ctx context.Context, _ ListOptions) ([]Inventory, error) {
+	objs, err := cic.client.List(ctx, metav1.ListOptions{})
+	if err != nil {
 		return nil, err
 	}
-	if apierrors.IsNotFound(err) {
-		return map[string]object.ObjMetadataSet{}, nil
-	}
-
-	identifiers := make(map[string]object.ObjMetadataSet)
-
-	for i, inv := range clusterInvs.Items {
-		invName := inv.GetName()
-		identifiers[invName] = object.ObjMetadataSet{}
-		invObj, err := cic.InventoryFactoryFunc(&clusterInvs.Items[i])
-		if err != nil {
-			return nil, err
-		}
-		wrappedInvObjSlice, err := invObj.Load()
+	var inventories []Inventory
+	for _, obj := range objs.Items {
+		uInv, err := cic.fromUnstructured(&obj)
 		if err != nil {
 			return nil, err
 		}
-		identifiers[invName] = append(identifiers[invName], wrappedInvObjSlice...)
+		inventories = append(inventories, uInv)
 	}
-
-	return identifiers, nil
+	return inventories, nil
 }
 
-// deleteInventoryObjByName deletes the passed inventory object from the APIServer, or
-// an error if one occurs.
-func (cic *ClusterClient) deleteInventoryObjByName(ctx context.Context, obj *unstructured.Unstructured, dryRun common.DryRunStrategy) error {
-	if obj == nil {
-		return fmt.Errorf("attempting delete a nil inventory object")
-	}
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run delete inventory object: not deleted")
+// CreateOrUpdate the in-cluster inventory
+// Updates the unstructured object, or creates it if it doesn't exist
+func (cic *UnstructuredClient) CreateOrUpdate(ctx context.Context, inv Inventory, _ UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
+	}
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
+	}
+	// Attempt to retry on a resource conflict error to avoid needing to retry the
+	// entire Apply/Destroy when there's a transient conflict.
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		create := false
+		obj, err := cic.client.Namespace(ui.Namespace()).Get(ctx, ui.Name(), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			create = true
+		} else if err != nil {
+			return err
+		}
+		if obj != nil { // Obj is nil when IsNotFound, in this case keep initial/empty obj
+			ui.ClusterObj = obj
```

Comment:
This will replace or erase the labels and annotations set on the Inventory by the Info...

I think we need to distinguish between the desired and current metadata.